go/vt/vtgate/vstream_manager.go

```diff
@@ -404,25 +418,66 @@ func (vs *vstream) streamFromTablet(ctx context.Context, sgtid *binlogdatapb.Sha
 
 		var eventss [][]*binlogdatapb.VEvent
 		var err error
-		rss := vs.rss
-		if vs.resolver != nil {
-			rss, err = vs.resolver.ResolveDestination(ctx, sgtid.Keyspace, vs.tabletType, key.DestinationShard(sgtid.Shard))
-			if err != nil {
-				return err
-			}
+		tp, err := discovery.NewTabletPicker(vs.ts, []string{vs.vsm.cell}, sgtid.Keyspace, sgtid.Shard, vs.tabletType.String())
+		if err != nil {
+			log.Errorf(err.Error())
+			return err
 		}
-		if len(rss) != 1 {
-			// Unreachable.
-			return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "unexpected number or shards: %v", rss)
+		tablet, err := tp.PickForStreaming(ctx)
+		if err != nil {
+			log.Errorf(err.Error())
+			return err
 		}
+		log.Infof("Picked tablet %s for for %s/%s/%s/%s", tablet.Alias.String(), vs.vsm.cell, sgtid.Keyspace, sgtid.Shard, vs.tabletType.String())
+		target := &querypb.Target{
+			Keyspace:   sgtid.Keyspace,
+			Shard:      sgtid.Shard,
+			TabletType: vs.tabletType,
+			Cell:       vs.vsm.cell,
+		}
+		tabletConn, err := vs.vsm.resolver.GetGateway().QueryServiceByAlias(tablet.Alias, target)
+		if err != nil {
+			log.Errorf(err.Error())
+			return err
+		}
+
+		errCh := make(chan string, 1)
+		go func() {
+			tabletConn.StreamHealth(ctx, func(shr *querypb.StreamHealthResponse) error {
```

Comment:
Just an idea: this seems something that is useful for all stream use cases. Would it make sense to move this logic inside `PickForStreaming` and change the interface of it to
```
PickForStreaming(ctx context.Context) (*topodatapb.Tablet, chan string, error)
```

Reply:
I get your point: for example vreplication streams should also track tablet changes. 

However the tablet picker is currently a passive module. Making it track the health checks of the tablets will add more complexity and break the usage pattern in other modules. I will keep this in mind when we update vreplication to use this logic and see if this can be wrapped meaningfully in the picker or elsewhere.